server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoverySourceService.java

```diff
@@ -218,6 +255,24 @@ synchronized RecoverySourceHandler addNewRecovery(StartRecoveryRequest request,
                 return handler;
             }
 
+            /**
+             * Adds recovery source handler.
+             */
+            synchronized void reestablishRecovery(ReestablishRecoveryRequest request, ActionListener<RecoveryResponse> listener) {
+                RecoverySourceHandler handler = null;
+                for (RecoverySourceHandler existingHandler : recoveryHandlers) {
+                    if (existingHandler.getRequest().recoveryId() == request.recoveryId()) {
```

Comment:
We also need to compare `targetAllocationId` as we can have two ongoing recoveries for the same shard with the same recoveryId from different targets.